src/sdam/description/topology/mod.rs

```diff
@@ -462,14 +464,48 @@ impl TopologyDescription {
     /// Update the topology based on the new information about the topology contained by the
     /// ServerDescription.
     pub(crate) fn update(&mut self, mut server_description: ServerDescription) -> Result<()> {
-        // Ignore updates from servers not currently in the cluster.
-        if !self.servers.contains_key(&server_description.address) {
-            return Ok(());
+        match self.servers.get(&server_description.address) {
+            None => return Ok(()),
+            Some(existing_sd) => {
+                // Ignore updates from outdated topology versions.
+                if let Some(existing_tv) = existing_sd.topology_version() {
+                    if let Some(new_tv) = server_description.topology_version() {
+                        if existing_tv.process_id == new_tv.process_id
+                            && new_tv.counter < existing_tv.counter
+                        {
+                            return Ok(());
+                        }
+                    }
+                }
+            }
         }
 
-        // Update the round trip time on the server description to the weighted average as described
-        // by the spec.
-        self.update_round_trip_time(&mut server_description);
+        if let Some(expected_name) = &self.set_name {
```

Comment:
This used to be checked in the Topology worker task, but it makes more sense to do it at the `TopologyDescription` level.